manipulate Queries via APIv3

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/21365
## Description

This PR allows to perform all CRUD operations on queries.
## Progress
- [x] list queries
  - [x] global ones (`GET /api/v3/queries`)
  - [x] of a project (`GET /api/v3/projects/:id/queries`)
- [ ] create query
  - [ ] globally (`POST /api/v3/queries`)
  - [ ] on a project (`POST /api/v3/projects/:id/queries`)
  - [ ] corresponding form endpoints (`POST /api/v3/queries/form`, `POST /api/v3/projects/:id/queries/form`)
- [ ] edit query
  - [x] PATCH endpoint  (`PATCH /api/v3/queries/:id`)
  - [ ] form endpoint (`POST /api/v3/queries/:id/form`)
- [x] delete query
- [ ] verify enforcement of permissions

NobodysNightmare committed:
Allow deletion of queries via APIv3

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-crud-queries-v3.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-crud-queries-v3.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
Move visibility check into a scope of query

NobodysNightmare committed:
Extract shared context from spec

NobodysNightmare committed:
Add list actions for queries

NobodysNightmare commented on spec/requests/api/v3/queries/query_resource_spec.rb:
```diff
@@ -56,9 +56,62 @@
+    context 'user not allowed' do
+      let(:permissions) { [:view_work_packages] }
+
+      it_behaves_like 'unauthorized access'
```
Note to self: check that the query was indeed not deleted...

NobodysNightmare committed:
Extend query delete spec

NobodysNightmare committed:
Extract serialization relevant methods into helper

NobodysNightmare committed:
Allow to edit queries via APIv3

NobodysNightmare committed:
Improve error messages for filter validation

NobodysNightmare commented on doc/apiv3-documentation.apib:
```diff
@@ -2028,6 +2028,126 @@ Lists the columns available for work packages of this project in the form of fie
                 "message": "The specified query does not exist."
             }
 
+## Edit Query [PATCH]
```
note to self: specify properties above as READ / WRITE

NobodysNightmare committed:
Rename Schema to SchemaRepresenter

for a better distinction between the represented value and its representer

NobodysNightmare commented:
To anyone that will continue working here:

The straightforward part should be implementing the remaining create/edit endpoints. You will basically only have to accept the same format that is returned by the API as input and flavour it with some error handling.

The interesting/hard/not-yet-defined part is how exactly the schema for queries will look like (which is part of the forms). This part is quite interesting, since the form is used to convey the available options to the client, that makes it somewhat interesting.

For the `groupBy` property it is quite straightforward to come up with a schema based on the currently exisiting schemas. In UI terms it reflects a simple single select, therefore it is not different from the work package priority. I.e. you can just use string objects in your allowed values:

``` JSON
"allowedValues": [
  { "href": "/api/v3/string_objects/subject" },
  { "href": "/api/v3/string_objects/type" },
  { "href": "/api/v3/string_objects/status" }
]
```

Many other properties of a query are special in some kind and there exists nothing comparable in the work package schema. Thus we have to think how those properties can be represented in our schema:
- `columns` are quite similar to `groupBy`. You have a set of allowed values, but as a twist you can select more than one and you are supposed to pass them as array. This needs to be conveyed to the client
- `sortCriteria` get more complicated, because you have a set of columns available for sorting and you are supposed to pass them along with a sort order as nested JSON arrays (e.g. `[["subject", "asc"], ["more columns", "desc"]]`)
- `filters` are quite complex JSON objects
  - you can add a filter for different columns
  - depending on the column there are other operators available
  - depending on column and operator only some values might be allowed
  - again you can have more than one

I have not yet made up a solution for most of those problems. It is probably okay to not specify allowed values for the most complex cases and just say "the client will have to know, it is too hard to explain". But at least for key questions there needs to be some way to find out, ideally using the existing means of a `schema`.

NobodysNightmare committed:
Extract base class out of form representer

- moving error handling and basic structure of a form into a generic base class
- leaving work package specific stuff in WP form representer

NobodysNightmare committed:
Add a representer for a query form and a basic schema

- schema does not yet indicate allowed values
- there are no endpoints yet to house the schema and form
- ensure that the QueryRepresenter will not return nil for filters/columns etc.
  (that is what we promise in the schema using `required: true`)

ulferts commented:
With #5233 being almost done, we can safely close this now.
